controllers/storageclaim_controller.go

```diff
@@ -465,6 +364,214 @@ func (r *StorageClaimReconciler) reconcilePhases() (reconcile.Result, error) {
+				// delete groupname from data as its not required in storageclass
+				delete(data, "subvolumegroupname")
+				storageClass = r.getCephFSStorageClass(data)
+			} else if resource.Name == "ceph-rbd" {
+				storageClass = r.getCephRBDStorageClass(data)
+			}
+			utils.AddAnnotation(storageClass, storageClaimAnnotation, r.storageClaim.Name)
+			err = r.createOrReplaceStorageClass(storageClass)
+			if err != nil {
+				return fmt.Errorf("failed to create or update StorageClass: %s", err)
+			}
+		case "VolumeSnapshotClass":
+			var volumeSnapshotClass *snapapi.VolumeSnapshotClass
+			data["csi.storage.k8s.io/snapshotter-secret-namespace"] = r.OperatorNamespace
+			// generate a new clusterID for cephfs subvolumegroup, as
+			// storageclaim is clusterscoped resources using its
+			// hash as the clusterID
+			data["clusterID"] = r.storageClaimHash
+			if resource.Name == "cephfs" {
+				volumeSnapshotClass = r.getCephFSVolumeSnapshotClass(data)
+			} else if resource.Name == "ceph-rbd" {
+				volumeSnapshotClass = r.getCephRBDVolumeSnapshotClass(data)
+			}
+			utils.AddAnnotation(volumeSnapshotClass, storageClaimAnnotation, r.storageClaim.Name)
+			if err := r.createOrReplaceVolumeSnapshotClass(volumeSnapshotClass); err != nil {
+				return fmt.Errorf("failed to create or update VolumeSnapshotClass: %s", err)
+			}
+		}
+	}
+
+	// update monitor configuration for cephcsi
+	err = cc.UpdateMonConfigMap(csiClusterConfigEntry.ClusterID, r.storageClient.Status.ConsumerID, csiClusterConfigEntry)
+	if err != nil {
+		return fmt.Errorf("failed to update mon configmap: %v", err)
+	}
+	return nil
+}
+
+func (r *StorageClaimReconciler) reconcileConfigResources(rpcClient *providerclient.OCSProviderClient) error {
+	response, err := rpcClient.GetStorageClaimConfig(
+		r.ctx,
+		r.storageClient.Status.ConsumerID,
+		r.storageClaim.Name,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to get StorageClaim config: %v", err)
+	}
+	resources := response.ExternalResource
+	if len(resources) == 0 {
+		return fmt.Errorf("no storageclaim configuration data received")
+	}
+	csiConfig := &csiopv1a1.Config{}
+	csiConfig.Name = r.storageClaimHash
+	csiConfig.Namespace = r.OperatorNamespace
+	for _, resource := range resources {
+		data := map[string]string{}
+		if err := json.Unmarshal(resource.Data, &data); err != nil {
+			return fmt.Errorf("failed to unmarshal StorageClaim configuration response: %v", err)
+		}
+		switch resource.Kind {
+		case "Secret":
+			secret := &corev1.Secret{}
+			secret.Name = resource.Name
+			secret.Namespace = r.OperatorNamespace
+			if _, err := controllerutil.CreateOrUpdate(r.ctx, r.Client, secret, func() error {
+				secret.Data = make(map[string][]byte, len(data))
+				for k, v := range data {
+					secret.Data[k] = []byte(v)
+				}
+				return r.own(secret)
+			}); err != nil {
+				return fmt.Errorf("failed to ensure secret %s: %v", secret.Name, err)
+			}
+		case "StorageClass":
```

Comment:
can we avoid this duplication and handle the config and CR part inside a if/else check which can minimise the code duplication?

Reply:
not possible at the moment, every where I'm using big branching rather than minor variants w/ li'l copying because at the end only one survives before going to QE.

There were some implicit assumptions in existing code on how the type of storageclass is found out, I don't see a reason to follow that.

If it's not mandatory, I'd want to defer this comment.

Reply:
done.